### PR TITLE
Adding date suggestions to the documents details view

### PR DESCRIPTION
**Pull request by Eckii24**
## Proposed change

Suggestions for tags, correspondent and the document type are made in the detail view of a document.
The value of the creation date is only pre-assigned once after parsing with the first date found in the document.

This change includes a suggestion list for other date values found in the document.

image

To achieve this, the parsing function is adjusted to be able to parse all date occurrences within the document.
Since this can be very time consuming for large documents, a setting is also introduced to control how many dates of the document should be output as a suggestion. If this setting is set to zero, Paperless behaves exactly as before.

An other approach to potentially improve the performance could be to use `dateparser.search.search_dates` like mentioned in https://github.com/paperless-ngx/paperless-ngx/discussions/741. ~~I have not tested this yet...~~
_EDIT: I quickly checked the usage of `search_dates`. Right now the implementation does not cover the common date formats of the current implementation. This may be improved when the PR there to re-implement the feature is merged...._

This addresses https://github.com/paperless-ngx/paperless-ngx/discussions/384

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
This is an awesome PR, thank you! Highly requested. I see some very small issues with frontend code but I will find some time to test and review this hopefully in the next few days.

**Review by shamoon — APPROVED** [review]
The frontend code looks good to me, I will let others review on backend. In general this works pretty well in testing, though I realize the accuracy of the suggested dates comes from a dependency.

One thought: I would be in favor of setting `PAPERLESS_NUMBER_OF_SUGGESTED_DATES` to 3 (or something sane), the other suggestions are enabled by default so I suspect people will not find / use this if its not. And frankly I think its super useful.

Thanks again for your contribution

File: src-ui/src/app/components/common/input/date/date.component.ts [hunk 1]
```diff
@@ -53,17 +53,11 @@ export class DateComponent
   }
 
   onSuggestionClick(dateString: string) {
-    const parsedNgDate = this.ngbDateParserFormatter.parse(dateString)
-    this.writeValue(this.formatDateAsYYYYMMDD(parsedNgDate))
+    const parsedDate = this.ngbDateParserFormatter.parse(dateString)
+    this.writeValue(`${parsedDate.year}-${parsedDate.month}-${parsedDate.day}`)
```
Comment (Eckii24) [inline src-ui/src/app/components/common/input/date/date.component.ts:57]: Hey @shamoon,

thanks for your review!

Did you test your change with a date like 1.1.22?
I had the problem, that the date input field expects a date with two digits for month and day, otherwise it did not work for me.

**Comment by Eckii24** [comment]
> One thought: I would be in favor of setting `PAPERLESS_NUMBER_OF_SUGGESTED_DATES` to 3 (or something sane), the other suggestions are enabled by default so I suspect people will not find / use this if its not. And frankly I think its super useful.

First of all I had a default of 10. The problem comes into place, if you work on very large documents. The performance of the date parsing library is not that good and the request to the suggestion api will take up to 10 sec. with is pretty noticeable.

Therefore my intention was to not introduce a kind of „breaking change“ in those cases.

That’s also the reason why I took the dates from the top to the bottom of the document and did not decide about the relevance of each date, because if you really parse all dates in the documents it took in my test for about 25 pages up to 30 sec…

File: src-ui/src/app/components/common/input/date/date.component.ts (hunk 1 above)
Comment (shamoon) [reply src-ui/src/app/components/common/input/date/date.component.ts]: Im not seeing that issue but there is a problem with dates being off by 1 day, that is due to using `Date()` which automatically uses timezone (see #957). I pushed another commit to deal with this. Can you test as-is and tell me if youre seeing any issues?

**Comment by shamoon** [comment]
Would love to hear others' thoughts on the `PAPERLESS_NUMBER_OF_SUGGESTED_DATES` question. In real world testing it hasn't slowed things down for me so I would be in favor of a default > 0, but totally understand your concerns.

**Comment by qcasey** [comment]
Thank you for this PR, it looks really useful!

The parsing time is very dependent on the document and hardware being used. With some of my testing documents the extra parsing takes:

### 23 pg without dates

- Laptop: ~2.3 seconds
- RPi 3A: ~55.4 seconds

### 176 pg with date on pg 160

- Laptop: ~0.2855 seconds
- RPi 3A: ~5.3 seconds

That said, I did not experience noticeable slowdowns when date parsing on either machine. 

> the request to the suggestion api will take up to 10 sec. with is pretty noticeable.

I'm curious where this is noticeable? The only difference I noticed was the date suggestions did not appear until they were parsed. If there is a hit to document load times, processing, etc. then I agree leaving it off should probably be the default. Right now the change seems benign to me.

File: src/documents/views.py [hunk 2]
```diff
@@ -329,13 +332,15 @@ def metadata(self, request, pk=None):
...
     def suggestions(self, request, pk=None):
-        try:
-            doc = Document.objects.get(pk=pk)
-        except Document.DoesNotExist:
-            raise Http404()
+        doc = get_object_or_404(Document, pk=pk)
 
         classifier = load_classifier()
 
+        gen = parse_date_generator(doc.filename, doc.content)
```
Comment (stumpylog) [inline src/documents/views.py:339]: So this parses the filename and/or content for each request?  Could this be cached in some way?  Or stored in the DB during the initial parse (if some upper limit is set)?

Comment (Eckii24) [reply src/documents/views.py:339]: I would say almost anything is possible. The question is whether it makes sense to cache the date values or store them in the database.

The usecase for these date suggestions is in my eyes the following: 
> A new document is imported and then the settings for that document are checked and adjusted if necessary.

In my eyes, this is a one-time task, which is why I don't think caching makes sense, since it won't take effect the first time it's loaded.

I see it similarly with the persistence of the suggestions in the database. Since this is data that is probably only needed once, I would not store something like this in the database.

What might be worth further discussion is server-side caching as part of the parsing (since this already takes some time anyway). This cache could then be retrieved and cleared the first time the document is loaded. This would have two advantages in my eyes:
* The waiting time to retrieve the suggestions when editing documents is significantly reduced.
* We can load all the document's date values and apply more complex logic for the selection of the suggestions

What do you think?

**Comment by Eckii24** [comment]
Hi @qcasey 

> I'm curious where this is noticeable? I suppose like shamoon, the only difference I noticed was the date suggestions did not appear until they were parsed. If there is a hit to document load times, processing, etc. then I agree leaving it off should probably be the default. Right now the change seems benign to me.

That is exactly the point I wanted to make. The time until the suggestions appear (not the time to load the other document values) is delayed depending on the document size and hardware.

If the complete suggestions take 10 seconds or longer, one or the other will go on without the date suggestions and continue to enter the value manually, but have the other suggestions directly at hand to use them.

In other words, if the suggestions take too long to load, the document editing process will be negatively affected.

A solution to this potential problem could be an improvement in the date parsing logic (but current I do not know how to improve it) or introduction of caching or something like that (see https://github.com/paperless-ngx/paperless-ngx/pull/1367#discussion_r941577694).

Do you see a problem here as well?

File: src/documents/views.py (hunk 2 above)
Comment (stumpylog) [reply src/documents/views.py:339]: I agree that once a user has edited the date or taken one of the suggestions, it wouldn't be needed anymore, so a persistent storage of located dates doesn't make sense.

Could the response be cached in the same way the thumbnail is, using Django's `cache_control`?  I'm not really a web expert.  That could be a benefit for the classifier loading as well being cached.

My ideal would be parsing most dates during the initial consumption, but I'm not sure how best to connect the consumption to the API in the best way.

**Comment by qcasey** [comment]
> one or the other will go on without the date suggestions and continue to enter the value manually, but have the other suggestions directly at hand to use them.

Please correct me if I'm wrong, but you're saying that if:

1. `PAPERLESS_NUMBER_OF_SUGGESTED_DATES=3` 
2. Paperless finds 2 dates somewhat quickly but takes 55+ seconds to find the 3rd date.

Then the user will not have any suggested dates to choose because the parsing of all 3 takes too long.

If I'm reading that right, would a default of `PAPERLESS_NUMBER_OF_SUGGESTED_DATES=1` still exhibit this issue? In this case, if the date suggestions take too long to load then there likely isn't a suggestion available anyway.

File: src-ui/src/app/components/common/input/date/date.component.ts (hunk 1 above)
Comment (Eckii24) [reply src-ui/src/app/components/common/input/date/date.component.ts]: I have tested your change. Looks good. I did not see any issues.

Thanks for improving the frontend :)

**Comment by stumpylog** [comment]
I've lost the thread on the status of this.  i know it's often requested, so it would be nice to make it available.  Even if defaulted to off for the moment while figuring out performance implications

**Comment by shamoon** [comment]
Agree we should move ahead, only question is about defaulting to off or on. I think we should default to on, maybe leave `PAPERLESS_NUMBER_OF_SUGGESTED_DATES` low, like 2? No one will see it otherwise. My feeling is if people have issues with the performance after release we're sure to hear about it =)

Unless there are obvious ways to speed things up, bug again, I'd be happy with as-is and potentially improving after-the-fact

**Comment by stumpylog** [comment]
Yeah, I saw we enable it, set to 2 or 3, and see what user's think about any speed issues.

**Comment by shamoon** [comment]
Great, I set it to 3. Any more issues with this PR?

**Review by qcasey — APPROVED** [review]

File: paperless.conf.example [hunk 3]
```diff
@@ -69,6 +69,7 @@
 #PAPERLESS_POST_CONSUME_SCRIPT=/path/to/an/arbitrary/script.sh
 #PAPERLESS_FILENAME_DATE_ORDER=YMD
 #PAPERLESS_FILENAME_PARSE_TRANSFORMS=[]
+#PAPERLESS_NUMBER_OF_SUGGESTED_DATES=5
```
[the remark this answers is not in the document]
Comment (shamoon) [reply paperless.conf.example:72]: Yea I wasn't sure what the best thing is, the default is now 3 so putting `=3` would actually have no effect... Should we just remove it from the example conf?

[the remark this answers is not in the document]
Comment (qcasey) [reply paperless.conf.example:72]: I was also thinking that, but most of the other configs in this file don't correspond to the defaults :shrug:

[the remark this answers is not in the document]
Comment (shamoon) [reply paperless.conf.example:72]: Yea I sorta read the conf.example as like "here's some stuff you can set to change things" but it is odd how we include optional things when it probably makes more sense to just refer to the docs which are more detailed anyway

**Comment by Eckii24** [comment]
A default of 3 sounds greats! Thanks for your feedback and the adjustments.

I hope I will find some time next week, because I want to try a caching mechanism. If this will lead to an improvement I will submit a new PR, with the caching in place.

So from my point of view, if there are no more other open issues, we can merge this one here.

**Comment by shamoon** [comment]
Caching would be great if you can figure it out. I think we're good to merge this, its going into `dev` anyway and we can test for next release 1.8.1. Thanks @Eckii24 !

**Comment by fantasticle** [comment]
I'm looking so forward to it! This will be an awesome feature! Maybe we can introduce a ML algorithm in the future that will autofill the "correct" date by learning from other documents falling in the same category (e.g.: docs 1-100 always use the second date that appears, doc 101 will also use the 2nd date).
